src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpHandlerResourceV3Provider.cs

```diff
@@ -105,17 +116,44 @@ public CredentialPromptWebRequestHandler(IWebProxy proxy, ICredentials credentia
                         if (ProxyAuthenticationRequired(ex) &&
                             HttpHandlerResourceV3.PromptForProxyCredentials != null)
                         {
-                            // prompt use for proxy credentials.
-                            var credentials = await HttpHandlerResourceV3
-                                .PromptForProxyCredentials(request.RequestUri, Proxy, cancellationToken);
+                            ICredentials currentCredentials = Proxy.Credentials;
 
-                            if (credentials == null)
+                            try
                             {
-                                throw;
+                                await _credentialPromptLock.WaitAsync();
+
+                                // Check if the credentials have already changed
+                                if (beforeAuthId != _lastAuthId)
+                                {
+                                    continue;
+                                }
+
+                                // Limit the number of retries
+                                _authRetries++;
+                                if (_authRetries >= MaxAuthRetries)
```

Comment:
Because there is no actual difference, and you end up with random retry numbers floating through the codebase.

Can you explain why you retry 10 for auth, but 3 or 5 for 503 (501 is not implemented, so I guess you will never retry it)?